AppDB/datastore_server.py

```diff
@@ -2294,19 +2294,57 @@ def __kind_query(self, query, filter_info, order_info):
     if startrow > endrow:
       return []
 
-    result = self.datastore_batch.range_query(dbconstants.APP_KIND_TABLE, 
-                                              dbconstants.APP_KIND_SCHEMA, 
-                                              startrow, 
-                                              endrow, 
-                                              limit, 
-                                              offset=0, 
-                                              start_inclusive=start_inclusive, 
-                                              end_inclusive=end_inclusive)
-
-    fetched_entities = self.__fetch_entities(result, clean_app_id(query.app()))
+    # Since the validity of each reference is not checked until after the
+    # range query has been performed, we may need to fetch additional
+    # references in order to satisfy the query.
+    entities = []
+    current_limit = limit
+    while True:
+      references = self.datastore_batch.range_query(
+        dbconstants.APP_KIND_TABLE,
+        dbconstants.APP_KIND_SCHEMA,
+        startrow,
+        endrow,
+        current_limit,
+        offset=0,
+        start_inclusive=start_inclusive,
+        end_inclusive=end_inclusive
+      )
+
+      new_entities = self.__fetch_entities(references, clean_app_id(query.app()))
+      entities.extend(new_entities)
+
+      # If we have enough valid entities to satisfy the query, we can stop
```

Comment:
I think the comment reads better with something like: If we satisfied the original query, we are done.